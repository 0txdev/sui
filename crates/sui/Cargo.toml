[package]
name = "sui"
version.workspace = true
authors = ["Mysten Labs <build@mystenlabs.com>"]
license = "Apache-2.0"
publish = false
edition = "2021"

[dependencies]
json_to_table.workspace = true
tabled.workspace = true
anemo.workspace = true
anyhow.workspace = true
serde.workspace = true
serde_json.workspace = true
serde_yaml.workspace = true
signature.workspace = true
camino.workspace = true
tokio = { workspace = true, features = ["full"] }
async-trait.workspace = true
tracing.workspace = true
bcs.workspace = true
clap.workspace = true
bip32.workspace = true
rand.workspace = true
tap.workspace = true
inquire.workspace = true
rusoto_core.workspace = true
rusoto_kms.workspace = true
prometheus.workspace = true
git-version.workspace = true
const-str.workspace = true
num-bigint.workspace = true
regex.workspace = true
reqwest.workspace = true
im.workspace = true

sui-config.workspace = true
sui-execution = { path = "../../sui-execution" }
sui-swarm-config.workspace = true
sui-genesis-builder.workspace = true
sui-types.workspace = true
sui-json.workspace = true
sui-swarm.workspace = true
sui-json-rpc-types.workspace = true
sui-sdk.workspace = true
sui-keys.workspace = true
sui-source-validation.workspace = true
sui-move = { workspace = true, features = ["all"] }
sui-move-build.workspace = true
sui-protocol-config.workspace = true
shared-crypto.workspace = true
sui-replay.workspace = true

fastcrypto.workspace = true
fastcrypto-zkp.workspace = true

rustyline.workspace = true
rustyline-derive.workspace = true
colored.workspace = true
unescape.workspace = true
shell-words.workspace = true

tempfile.workspace = true
telemetry-subscribers.workspace = true

move-core-types.workspace = true
move-package.workspace = true
csv.workspace = true
workspace-hack.workspace = true

[target.'cfg(not(target_env = "msvc"))'.dependencies]
jemalloc-ctl.workspace = true

move-vm-test-utils = {workspace = true, optional = true}
move-vm-types = {workspace = true, optional = true}
move-vm-profiler = {workspace = true, optional = true}
move-cli = {workspace = true, optional = true}
move-unit-test = {workspace = true, optional = true}
move-vm-config = {workspace = true, optional = true}
sui-execution = {workspace = true, optional = true}
#move-vm-runtime = { path = "../../external-crates/move/move-execution/next-vm/crates/move-vm-runtime", option = true}

[dev-dependencies]
prometheus.workspace = true
fs_extra.workspace = true
expect-test.workspace = true
assert_cmd.workspace = true

test-cluster.workspace = true
sui-macros.workspace = true
sui-simulator.workspace = true
sui-test-transaction-builder.workspace = true

[package.metadata.cargo-udeps.ignore]
normal = ["jemalloc-ctl"]

[[example]]
name = "generate-genesis-checkpoint"
path = "src/generate_genesis_checkpoint.rs"
test = false

[[bin]]
name = "sui"
path = "src/main.rs"

[features]
default = ["gas-profiler"]
gas-profiler = [
    "sui-types/gas-profiler",
    "move-cli/gas-profiler",
    "move-unit-test/gas-profiler",
    "move-vm-config/gas-profiler",
    "move-vm-profiler/gas-profiler",
    "move-vm-types/gas-profiler",
    "move-vm-test-utils/gas-profiler",
    "sui-execution/gas-profiler",
    #"move-vm-runtime/gas-profiler",
]
