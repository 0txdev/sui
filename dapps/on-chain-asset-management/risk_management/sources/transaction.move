module risk_management::transaction {

    use risk_management::policy_config::{Self, SpenderCap, ApproverCap, Assets};
    use sui::tx_context::{Self, TxContext};
    use sui::object::{Self, UID, ID};
    use sui::coin::{Self};
    use std::string::{Self, String};
    use sui::transfer;
    use std::vector as vec;

    const EAmountLimitExceeded : u64 = 0;
    const ETimeLimitExceeded : u64 = 1;
    const EAlreadyApprovedByThisApprover : u64 = 2;
    const ENotOriginalOwnerOfCapability : u64 = 3;
    const ESpenderCannotSendMoneyToHimself : u64 = 4;

    struct TransactionRequest has key, store {
        id: UID,
        amount: u64,
        time_limit: u64,
        spender: address,
        recipient: address,
        description: String,
        approvers_num: u64,
        approved_by: vector<address>,
    }

    struct TransactionApproval has key, store {
        id: UID,
        transaction_id: ID,
        amount: u64,
        spender: address,
        recipient: address,
        approvers: vector<address>,
    }

    entry fun initiate_transaction(
        spender_cap: &SpenderCap,
        amount: u64,
        recipient: address,
        description: vector<u8>,
        ctx: &mut TxContext,
    ) { 
        assert!(policy_config::get_spender_original_owner(spender_cap) == tx_context::sender(ctx), 3);
        assert!(recipient != tx_context::sender(ctx), 4);
        assert!(amount <= policy_config::get_amount_limit(spender_cap), 0);
        transfer::share_object(TransactionRequest {
            id: object::new(ctx),
            amount,
            time_limit: policy_config::get_time_limit(spender_cap) + tx_context::epoch(ctx),
            spender: tx_context::sender(ctx),
            recipient,
            description: string::utf8(description),
            approvers_num: policy_config::get_approvers_num(spender_cap),
            approved_by: vec::empty()
        })
    }

    entry fun approve_request(
        approver_cap: &ApproverCap,
        tx_request: &mut TransactionRequest,
        ctx: &mut TxContext,
    ) {
        assert!(policy_config::get_approver_original_owner(approver_cap) == tx_context::sender(ctx), 3);
        assert!(tx_context::epoch(ctx) <= tx_request.time_limit, 1);
        assert!(vec::contains(&tx_request.approved_by, &tx_context::sender(ctx)) == false, 2);
        if (vec::length(&tx_request.approved_by) < tx_request.approvers_num - 1) {
            vec::push_back(&mut tx_request.approved_by, tx_context::sender(ctx))
        } else {
            vec::push_back(&mut tx_request.approved_by, tx_context::sender(ctx));
            let newvector : vector<address> = tx_request.approved_by;
            transfer::transfer(TransactionApproval{
                id: object::new(ctx),
                transaction_id: object::uid_to_inner(&tx_request.id),
                amount: tx_request.amount,
                spender: tx_request.spender,
                recipient: tx_request.recipient,
                approvers: newvector,
            }, tx_request.spender);
        }
    }

    entry fun reject_request(
        _: &ApproverCap,
        _tx_request: &TransactionRequest,
        _ctx: &mut TxContext,
    ) {
        // An event should be emitted informing that the request with
        // transaction_id: object::uid_to_inner(&tx_request.id) got rejected
    }

    entry fun execute_transaction(
        spender_cap: &mut SpenderCap,
        tx_approval: TransactionApproval,
        assets: &mut Assets,
        ctx: &mut TxContext,
    ) {
        assert!(policy_config::get_spender_original_owner(spender_cap) == tx_context::sender(ctx), 3);
        assert!(tx_approval.amount <= policy_config::get_amount_limit(spender_cap), 0);
        transfer::transfer(
            coin::take(policy_config::get_foundation_balance_mut(assets), tx_approval.amount, ctx) , 
            tx_approval.recipient
        );
        
        //Unpack and delete the TransactionApproval
        let TransactionApproval { 
            id,
            transaction_id: _,
            amount: _,
            spender: _,
            recipient: _,
            approvers: _ } = tx_approval;
        object::delete(id)
    }

}