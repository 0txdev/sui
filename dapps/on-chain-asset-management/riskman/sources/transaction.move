module riskman::transaction {

    use riskman::policy_config::{Self, SpenderCap, ApproverCap, Assets};
    use sui::tx_context::{Self, TxContext};
    use sui::object::{Self, UID, ID};
    use sui::coin::{Self};
    use std::string::{Self, String};
    use sui::transfer;

    const EAmountLimitExceeded : u64 = 0;
    const ETimeLimitExceeded : u64 = 1;
    const EAmountLimitExceededAfterApproval : u64 = 2;

    struct TransactionRequest has key, store {
        id: UID,
        amount: u64,
        time_limit: u64,
        spender: address,
        recipient: address,
        description: String,
    }

    struct TransactionApproval has key, store {
        id: UID,
        transaction_id: ID,
        amount: u64,
        spender: address,
        recipient: address,
        approver: address,
    }

    entry fun initiate_transaction(
        spender_cap: &SpenderCap,
        amount: u64,
        recipient: address,
        description: vector<u8>,
        ctx: &mut TxContext,
    ) { 
        assert!(amount <= policy_config::get_amount_limit(spender_cap) - policy_config::get_amount_spent(spender_cap), 0);
        transfer::share_object(TransactionRequest {
            id: object::new(ctx),
            amount,
            time_limit: policy_config::get_time_limit(spender_cap) + tx_context::epoch(ctx),
            spender: tx_context::sender(ctx),
            recipient,
            description: string::utf8(description)
        })
    }

    entry fun approve_request(
        _: &ApproverCap,
        tx_request: &TransactionRequest,
        ctx: &mut TxContext,
    ) {
        assert!(tx_context::epoch(ctx) <= tx_request.time_limit, 1);
        transfer::transfer(TransactionApproval{
            id: object::new(ctx),
            transaction_id: object::uid_to_inner(&tx_request.id),
            amount: tx_request.amount,
            spender: tx_request.spender,
            recipient: tx_request.recipient,
            approver: tx_context::sender(ctx)
        }, tx_request.spender);
    }

    entry fun reject_request(
        _: &ApproverCap,
        _tx_request: &TransactionRequest,
        _ctx: &mut TxContext,
    ) {
        // An event should be emitted informing that the request with
        // transaction_id: object::uid_to_inner(&tx_request.id) got rejected
    }

    entry fun execute_transaction(
        spender_cap: &mut SpenderCap,
        tx_approval: TransactionApproval,
        assets: &mut Assets,
        ctx: &mut TxContext,
    ) {
        assert!(tx_approval.amount <= policy_config::get_amount_limit(spender_cap) - policy_config::get_amount_spent(spender_cap), 2);
        transfer::transfer(
            coin::take(policy_config::get_foundation_balance_mut(assets), tx_approval.amount, ctx) , 
            tx_approval.recipient
        );
        policy_config::update_spent(spender_cap, tx_approval.amount);
        
        //Unpack and delete the TransactionApproval
        let TransactionApproval { 
            id,
            transaction_id: _,
            amount: _,
            spender: _,
            recipient: _,
            approver: _ } = tx_approval;
        object::delete(id)
    }

}