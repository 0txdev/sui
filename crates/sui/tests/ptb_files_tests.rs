// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use std::{collections::BTreeMap, path::Path};
use sui::client_ptb::{ptb::PTB, ptb_builder::errors::render_errors};
use sui_types::transaction::{CallArg, ObjectArg};
use test_cluster::TestClusterBuilder;

const TEST_DIR: &str = "tests";

#[cfg(not(msim))]
#[tokio::main]
async fn test_ptb_files(path: &Path) -> Result<(), Box<dyn std::error::Error>> {
    let _ = miette::set_hook(Box::new(|_| {
        Box::new(
            miette::MietteHandlerOpts::new()
                .color(false)
                .width(80)
                .build(),
        )
    }));

    let fname = || path.file_name().unwrap().to_string_lossy().to_string();
    let file = path.to_str().unwrap().to_string();
    let args = vec!["--file".to_string(), file, "--preview".to_string()];
    let cwd = std::env::current_dir().unwrap();
    let mut commands = Vec::new();
    let ptb = PTB { args: args.clone() };
    ptb.parse_args(cwd, args, &mut BTreeMap::new(), &mut commands)?;

    // === PREVIEW ===
    let ptb_preview = ptb.preview(&commands.clone());
    let mut results = vec![];
    let preview_string = if let Some(ptb_preview) = ptb_preview {
        ptb_preview.to_string()
    } else {
        "".to_string()
    };
    results.push(" === PREVIEW === ".to_string());
    results.push(preview_string);

    // === PARSE COMMANDS ===
    let parsed_ptb_commands = ptb.parse_ptb_commands(commands.clone());
    match parsed_ptb_commands {
        Ok(ref commands) => commands,
        Err(e) => {
            // insta::assert_display_snapshot!(
            //     fname(),
            //     format!("=== FILE EXPANSION ERROR === \n {:?}", e)
            // );
            return Ok(());
        }
    };
    let parsed_ptb_commands = match parsed_ptb_commands {
        Ok(parsed) => parsed,
        Err(errors) => {
            let rendered = render_errors(&commands, errors);
            results.push(" === ERRORS AFTER PARSING INPUT COMMANDS === ".to_string());
            for e in rendered.iter() {
                results.push(format!("{:?}", e));
            }
            insta::assert_display_snapshot!(fname(), results.join("\n"));
            return Ok(());
        }
    };

    results.push(" === PARSED INPUT COMMANDS === ".to_string());

    for (_, c) in &parsed_ptb_commands {
        results.push(format!("cmd: {:?}", c));
    }

    // === BUILD PTB ===
    let test_cluster = TestClusterBuilder::new().build().await;
    let context = test_cluster.wallet;
    let client = context.get_client().await?;

    let built_ptb = ptb
        .parse_and_build_ptb(parsed_ptb_commands, &context, client)
        .await;

    if let Ok(ref ptb) = built_ptb {
        results.push(" === BUILT PTB === ".to_string());
        for (i, ca) in ptb.0.inputs.iter().enumerate() {
            results.push(format!("Input {}: {}", i, stable_call_arg_display(ca)));
        }
        for (i, c) in ptb.0.commands.iter().enumerate() {
            results.push(format!("Command {}: {}", i, c));
        }
    }

    // === BUILDING PTB ERRORS ===
    if let Err(e) = built_ptb {
        let rendered = render_errors(&commands, e);

        results.push(" === BUILDING PTB ERRORS === ".to_string());
        for e in rendered.iter() {
            results.push(format!("{:?}", e));
        }
    }

    // === FINALLY DO THE ASSERTION ===
    insta::assert_display_snapshot!(fname(), results.join("\n"));

    Ok(())
}

fn stable_call_arg_display(ca: &CallArg) -> String {
    match ca {
        CallArg::Pure(v) => format!("Pure({:?})", v),
        CallArg::Object(oa) => match oa {
            ObjectArg::ImmOrOwnedObject(_) => "ImmutableOrOwnedObject".to_string(),
            ObjectArg::SharedObject { mutable, .. } => {
                format!("SharedObject(mutable: {})", mutable)
            }
            ObjectArg::Receiving(_) => "Receiving".to_string(),
        },
    }
}

datatest_stable::harness!(test_ptb_files, TEST_DIR, r".*\.ptb$",);
