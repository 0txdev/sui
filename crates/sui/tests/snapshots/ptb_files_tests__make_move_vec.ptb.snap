---
source: crates/sui/tests/ptb_files_tests.rs
expression: "results.join(\"\\n\")"
---
 === PREVIEW === 
╭───────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ PTB Preview                                                                                                       │
├───────────────┬─────────────────────────────────────────────────┬─────────────────────────────────────────────────┤
│ command       │ from                                            │ value(s)                                        │
├───────────────┼─────────────────────────────────────────────────┼─────────────────────────────────────────────────┤
│ gas_budget    │ tests/ptb_files/make_move_vec/make_move_vec.ptb │ 1000                                            │
│ make_move_vec │ tests/ptb_files/make_move_vec/make_move_vec.ptb │ <u64> [1, 2,3, 4]                               │
│ make_move_vec │ tests/ptb_files/make_move_vec/make_move_vec.ptb │ <u64> []                                        │
│ make_move_vec │ tests/ptb_files/make_move_vec/make_move_vec.ptb │ <0x1::a::b> [@0x1,@0x2]                         │
│ make_move_vec │ tests/ptb_files/make_move_vec/make_move_vec.ptb │ <address> [@0x1,@0x2]                           │
│ make_move_vec │ tests/ptb_files/make_move_vec/make_move_vec.ptb │ <vector<address>> [vector[@0x1],vector[@0x2]]   │
│ make_move_vec │ tests/ptb_files/make_move_vec/make_move_vec.ptb │ <std::option::Option<u64>> [none,none]          │
│ make_move_vec │ tests/ptb_files/make_move_vec/make_move_vec.ptb │ <std::option::Option<u64>> [none,none]          │
│ make_move_vec │ tests/ptb_files/make_move_vec/make_move_vec.ptb │ <std::option::Option<u64>> [none,some(1)]       │
│ make_move_vec │ tests/ptb_files/make_move_vec/make_move_vec.ptb │ <std::option::Option<u64>> [none,some(some(1))] │
│ make_move_vec │ tests/ptb_files/make_move_vec/make_move_vec.ptb │ <sui::coin::Coin<sui::sui::SUI>> [gas]          │
│ preview       │ console                                         │ true                                            │
╰───────────────┴─────────────────────────────────────────────────┴─────────────────────────────────────────────────╯
 === PARSED INPUT COMMANDS === 
cmd: GasBudget(Spanned { span: Span { start: 0, end: 4, arg_idx: 0, file_scope: FileScope { file_command_index: 0, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: 1000 })
cmd: MakeMoveVec(Spanned { span: Span { start: 0, end: 5, arg_idx: 0, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: U64 }, Spanned { span: Span { start: 0, end: 11, arg_idx: 1, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: [Spanned { span: Span { start: 1, end: 2, arg_idx: 1, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: U64(1) }, Spanned { span: Span { start: 4, end: 5, arg_idx: 1, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: U64(2) }, Spanned { span: Span { start: 6, end: 7, arg_idx: 1, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: U64(3) }, Spanned { span: Span { start: 9, end: 10, arg_idx: 1, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: U64(4) }] })
cmd: MakeMoveVec(Spanned { span: Span { start: 0, end: 5, arg_idx: 0, file_scope: FileScope { file_command_index: 2, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: U64 }, Spanned { span: Span { start: 0, end: 2, arg_idx: 1, file_scope: FileScope { file_command_index: 2, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: [] })
cmd: MakeMoveVec(Spanned { span: Span { start: 0, end: 11, arg_idx: 0, file_scope: FileScope { file_command_index: 3, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Struct(ParsedStructType { fq_name: ParsedFqName { module: ParsedModuleId { address: Numerical(0x1), name: "a" }, name: "b" }, type_args: [] }) }, Spanned { span: Span { start: 0, end: 11, arg_idx: 1, file_scope: FileScope { file_command_index: 3, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: [Spanned { span: Span { start: 2, end: 5, arg_idx: 1, file_scope: FileScope { file_command_index: 3, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Address(0x1) }, Spanned { span: Span { start: 7, end: 10, arg_idx: 1, file_scope: FileScope { file_command_index: 3, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Address(0x2) }] })
cmd: MakeMoveVec(Spanned { span: Span { start: 0, end: 9, arg_idx: 0, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Address }, Spanned { span: Span { start: 0, end: 11, arg_idx: 1, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: [Spanned { span: Span { start: 2, end: 5, arg_idx: 1, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Address(0x1) }, Spanned { span: Span { start: 7, end: 10, arg_idx: 1, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Address(0x2) }] })
cmd: MakeMoveVec(Spanned { span: Span { start: 0, end: 17, arg_idx: 0, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Vector(Address) }, Spanned { span: Span { start: 0, end: 27, arg_idx: 1, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: [Spanned { span: Span { start: 7, end: 13, arg_idx: 1, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Vector([Spanned { span: Span { start: 9, end: 12, arg_idx: 1, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Address(0x1) }]) }, Spanned { span: Span { start: 20, end: 26, arg_idx: 1, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Vector([Spanned { span: Span { start: 22, end: 25, arg_idx: 1, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Address(0x2) }]) }] })
cmd: MakeMoveVec(Spanned { span: Span { start: 0, end: 26, arg_idx: 0, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Struct(ParsedStructType { fq_name: ParsedFqName { module: ParsedModuleId { address: Named("std"), name: "option" }, name: "Option" }, type_args: [U64] }) }, Spanned { span: Span { start: 0, end: 11, arg_idx: 1, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: [Spanned { span: Span { start: 1, end: 5, arg_idx: 1, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Option(Spanned { span: Span { start: 1, end: 5, arg_idx: 1, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: None }) }, Spanned { span: Span { start: 6, end: 10, arg_idx: 1, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Option(Spanned { span: Span { start: 6, end: 10, arg_idx: 1, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: None }) }] })
cmd: MakeMoveVec(Spanned { span: Span { start: 0, end: 26, arg_idx: 0, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Struct(ParsedStructType { fq_name: ParsedFqName { module: ParsedModuleId { address: Named("std"), name: "option" }, name: "Option" }, type_args: [U64] }) }, Spanned { span: Span { start: 0, end: 11, arg_idx: 1, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: [Spanned { span: Span { start: 1, end: 5, arg_idx: 1, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Option(Spanned { span: Span { start: 1, end: 5, arg_idx: 1, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: None }) }, Spanned { span: Span { start: 6, end: 10, arg_idx: 1, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Option(Spanned { span: Span { start: 6, end: 10, arg_idx: 1, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: None }) }] })
cmd: MakeMoveVec(Spanned { span: Span { start: 0, end: 26, arg_idx: 0, file_scope: FileScope { file_command_index: 8, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Struct(ParsedStructType { fq_name: ParsedFqName { module: ParsedModuleId { address: Named("std"), name: "option" }, name: "Option" }, type_args: [U64] }) }, Spanned { span: Span { start: 0, end: 14, arg_idx: 1, file_scope: FileScope { file_command_index: 8, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: [Spanned { span: Span { start: 1, end: 5, arg_idx: 1, file_scope: FileScope { file_command_index: 8, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Option(Spanned { span: Span { start: 1, end: 5, arg_idx: 1, file_scope: FileScope { file_command_index: 8, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: None }) }, Spanned { span: Span { start: 6, end: 13, arg_idx: 1, file_scope: FileScope { file_command_index: 8, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Option(Spanned { span: Span { start: 11, end: 12, arg_idx: 1, file_scope: FileScope { file_command_index: 8, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Some(U64(1)) }) }] })
cmd: MakeMoveVec(Spanned { span: Span { start: 0, end: 26, arg_idx: 0, file_scope: FileScope { file_command_index: 9, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Struct(ParsedStructType { fq_name: ParsedFqName { module: ParsedModuleId { address: Named("std"), name: "option" }, name: "Option" }, type_args: [U64] }) }, Spanned { span: Span { start: 0, end: 20, arg_idx: 1, file_scope: FileScope { file_command_index: 9, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: [Spanned { span: Span { start: 1, end: 5, arg_idx: 1, file_scope: FileScope { file_command_index: 9, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Option(Spanned { span: Span { start: 1, end: 5, arg_idx: 1, file_scope: FileScope { file_command_index: 9, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: None }) }, Spanned { span: Span { start: 6, end: 19, arg_idx: 1, file_scope: FileScope { file_command_index: 9, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Option(Spanned { span: Span { start: 11, end: 18, arg_idx: 1, file_scope: FileScope { file_command_index: 9, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Some(Option(Spanned { span: Span { start: 16, end: 17, arg_idx: 1, file_scope: FileScope { file_command_index: 9, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Some(U64(1)) })) }) }] })
cmd: MakeMoveVec(Spanned { span: Span { start: 0, end: 32, arg_idx: 0, file_scope: FileScope { file_command_index: 10, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Struct(ParsedStructType { fq_name: ParsedFqName { module: ParsedModuleId { address: Named("sui"), name: "coin" }, name: "Coin" }, type_args: [Struct(ParsedStructType { fq_name: ParsedFqName { module: ParsedModuleId { address: Named("sui"), name: "sui" }, name: "SUI" }, type_args: [] })] }) }, Spanned { span: Span { start: 0, end: 5, arg_idx: 1, file_scope: FileScope { file_command_index: 10, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: [Spanned { span: Span { start: 1, end: 4, arg_idx: 1, file_scope: FileScope { file_command_index: 10, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Gas }] })
cmd: Preview(Spanned { span: Span { start: 0, end: 5, arg_idx: 0, file_scope: FileScope { file_command_index: 11, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Bool(false) })
cmd: WarnShadows(Spanned { span: Span { start: 0, end: 5, arg_idx: 0, file_scope: FileScope { file_command_index: 12, name: "tests/ptb_files/make_move_vec/make_move_vec.ptb", name_index: 0 } }, value: Bool(false) })
cmd: Preview(Spanned { span: Span { start: 0, end: 4, arg_idx: 0, file_scope: FileScope { file_command_index: 1, name: "console", name_index: 0 } }, value: Bool(true) })
cmd: WarnShadows(Spanned { span: Span { start: 0, end: 5, arg_idx: 0, file_scope: FileScope { file_command_index: 2, name: "console", name_index: 0 } }, value: Bool(false) })
 === BUILT PTB === 
Input 0: Pure([1, 0, 0, 0, 0, 0, 0, 0])
Input 1: Pure([2, 0, 0, 0, 0, 0, 0, 0])
Input 2: Pure([3, 0, 0, 0, 0, 0, 0, 0])
Input 3: Pure([4, 0, 0, 0, 0, 0, 0, 0])
Input 4: ImmutableOrOwnedObject
Input 5: ImmutableOrOwnedObject
Input 6: Pure([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1])
Input 7: Pure([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2])
Input 8: Pure([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1])
Input 9: Pure([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2])
Input 10: Pure([0])
Input 11: Pure([1, 1, 0, 0, 0, 0, 0, 0, 0])
Input 12: Pure([1, 1, 1, 0, 0, 0, 0, 0, 0, 0])
Command 0: MakeMoveVec(Someu64,[Input(0),Input(1),Input(2),Input(3)])
Command 1: MakeMoveVec(Someu64,[])
Command 2: MakeMoveVec(Some0x1::a::b,[Input(4),Input(5)])
Command 3: MakeMoveVec(Someaddress,[Input(6),Input(7)])
Command 4: MakeMoveVec(Somevector<address>,[Input(8),Input(9)])
Command 5: MakeMoveVec(Some0x1::option::Option<u64>,[Input(10),Input(10)])
Command 6: MakeMoveVec(Some0x1::option::Option<u64>,[Input(10),Input(10)])
Command 7: MakeMoveVec(Some0x1::option::Option<u64>,[Input(10),Input(11)])
Command 8: MakeMoveVec(Some0x1::option::Option<u64>,[Input(10),Input(12)])
Command 9: MakeMoveVec(Some0x2::coin::Coin<0x2::sui::SUI>,[GasCoin])
